### PR TITLE
Better code blocks

### PR DESCRIPTION
**Pull request by jonludlam**
This commit adds some new functionality to code blocks. Firstly it allows arbitrary delimiters meaning that code containing the usual code block terminator can now be parsed correctly. The syntax for this is:

    {delimiter@ocaml[ ... ]delimiter}

The delimiter can contain the chars `[ a-z A-Z 0-9 _ - ]`, the same as the language tag that comes after the '@' symbol. Note that there's no way to have a delimited code block without a language tag.

The second piece of functionality is that code blocks can now have associated output:

    {@ocaml[

    ... ocaml code ...

    ][

    ... odoc formatted output ...

    ]}

This syntax also supports the delimiters as above. The delimiters only encode the _code_ block, not the output:

    {delim@ocaml[

    ... ocaml code containing ][ or ]} ...

    ]delim[

    ... odoc formatted output ...

    ]}

The idea is that the odoc formatted output should be well formed and thus any escaping is done in the usual way.

The output can then contain, for example, error blocks produced by mdx:

    {delim@ocaml[
      let x = "]}
    ]delim[
    {err@mdx-error[
    Line 1, characters 9-10:
    Error: String literal not terminated]err}
    ]}

In addition this also allows the possibility of code blocks to produce rich output - ie., allowing marked-up such as tables, headings, images and so on, in such a way that they are associated with the code block, and hence can be manipulated by a 'test-promote' workflow.

### CODE REVIEW

**Comment by jonludlam** [comment]
Review comment from group review in Cambridge: Could we perhaps make the delimiting generic, e.g.:

```
{delim@ocaml[ ... ]delim[ ]}
{delim|v ... verbatim v|delim}
{delim|m ... |delim}
{delim|math ... |delim}
{|v ... v|}
```

**Comment by jonludlam** [comment]
While I think having delimiters elsewhere is a worthy goal, we can do that in another PR.

File: src/syntax.ml [hunk 1]
```diff
@@ -671,7 +674,15 @@ let rec block_element_list :
...
     (* Terminators: the two tokens that terminate anything. *)
-    | ({ value = `End; _ } | { value = `Right_brace; _ }) as next_token -> (
+    | { value = `End; _ } as next_token -> (
+        match context with
+        | Top_level -> (List.rev acc, next_token, where_in_line)
+        | In_shorthand_list -> (List.rev acc, next_token, where_in_line)
+        | In_explicit_list -> (List.rev acc, next_token, where_in_line)
+        | In_tag -> (List.rev acc, next_token, where_in_line)
+        | In_table_cell -> (List.rev acc, next_token, where_in_line)
+        | In_code_results -> (List.rev acc, next_token, where_in_line))
```
Comment (Julow) [inline src/syntax.ml:684]: Most cases have the same type and could perhaps be written with an or-pattern ?

File: src/token.ml [hunk 2]
```diff
@@ -149,6 +152,7 @@ let describe : [< t | `Comment ] -> string = function
   | `Single_newline _ -> "line break"
   | `Blank_line _ -> "blank line"
   | `Right_brace -> "'}'"
+  | `Right_code_delimiter -> "']@}'"
```
Comment (Julow) [inline src/token.ml]: There's no such token.

File: src/lexer.mll [hunk 3]
```diff
@@ -675,11 +673,29 @@ and code_block_metadata_tail input = parse
...
-and code_block start_offset metadata input = parse
-  | (code_block_text as c) "]}"
-    { emit_code_block ~start_offset input metadata c }
-  | (code_block_text as c) eof
+and code_block start_offset content_offset metadata prefix delim input = parse
+  | ("]" (language_tag_char* as delim') "[") as terminator
+    { if delim = delim'
+      then emit_code_block ~start_offset content_offset input metadata delim terminator prefix true
+      else
+        (Buffer.add_string prefix ("]" ^ delim' ^ "[");
```
Comment (Julow) [inline src/lexer.mll]: That's `terminator`.

File: src/lexer.mll [hunk 4]
```diff
@@ -675,11 +673,29 @@ and code_block_metadata_tail input = parse
   | eof
     { `Eof }
 
-and code_block start_offset metadata input = parse
-  | (code_block_text as c) "]}"
-    { emit_code_block ~start_offset input metadata c }
-  | (code_block_text as c) eof
+and code_block start_offset content_offset metadata prefix delim input = parse
+  | ("]" (language_tag_char* as delim') "[") as terminator
```
Comment (Julow) [inline src/lexer.mll]: We should have a `delim_char`. Perhaps without `-` because `{-` exsists.

File: src/lexer.mll [hunk 5]
```diff
@@ -219,18 +217,19 @@ let emit_verbatim input start_offset buffer =
   let t = trim_trailing_blank_lines t in
   emit input (`Verbatim t) ~start_offset
 
-let emit_code_block ~start_offset input metadata c =
-  let c = trim_trailing_blank_lines c in
+let emit_code_block ~start_offset content_offset input metadata delim terminator c has_results =
```
Comment (Julow) [inline src/lexer.mll:226]: The location calculations in this function are quite complicated and would deserve some comments.
Why is `content_offset` needed ?

**Review by Julow — COMMENTED** [review]
The syntax looks reasonable, no need to block for more generic delimiters (which would probably still look different on code blocks).

File: src/lexer.mll (hunk 3 above)
Comment (jonludlam) [reply src/lexer.mll]: ah yep!

File: src/lexer.mll (hunk 4 above)
Comment (jonludlam) [reply src/lexer.mll]: Good plan

File: src/token.ml (hunk 2 above)
Comment (jonludlam) [reply src/token.ml]: Fixed

File: src/syntax.ml (hunk 1 above)
Comment (jonludlam) [reply src/syntax.ml:684]: Annoyingly not (the return types are different in each case). See the comment immediately below.

File: src/lexer.mll (hunk 5 above)
Comment (jonludlam) [reply src/lexer.mll:226]: I've added a comment.

File: src/syntax.ml (hunk 1 above)
Comment (Julow) [reply src/syntax.ml:684]: Indeed.

File: src/lexer.mll [hunk 6]
```diff
@@ -267,6 +267,9 @@ let raw_markup_target =
 let language_tag_char =
   ['a'-'z' 'A'-'Z' '0'-'9' '_' '-' ]
 
+let delim_char =
+  ['a'-'z' 'A'-'Z' '0'-'9' '_' '-' ]
```
Comment (Julow) [inline src/lexer.mll]: Without `-` ? This might interfere with the `{- ...}` syntax with a different parser engine.

File: src/lexer.mll (hunk 5 above)
Comment (Julow) [reply src/lexer.mll:226]: Thanks!

**Review by Julow — APPROVED** [review]
I have a remark on `delim_char` but otherwise I think it's ready to merge.
The CI failures are not related, everything is OK.

File: src/lexer.mll (hunk 6 above)
Comment (jonludlam) [reply src/lexer.mll]: Huh, I thought I had removed that, as per your previous comment. Will fix, thanks!

**Pull request merged**